UefiPayloadPkg/AcpiPlatformDxe/AcpiPlatform.c

```diff
@@ -74,7 +74,10 @@ InstallTablesFromXsdt (
       FadtTable = (EFI_ACPI_6_3_FIXED_ACPI_DESCRIPTION_TABLE *)
                     (UINTN) CurrentTablePointer;
       DsdtTable  = (EFI_ACPI_DESCRIPTION_HEADER *) (UINTN) FadtTable->XDsdt;
+
       FacsTable = (EFI_ACPI_6_3_FIRMWARE_ACPI_CONTROL_STRUCTURE *) (UINTN) FadtTable->XFirmwareCtrl;
+      if (FacsTable == NULL)
+          FacsTable = (EFI_ACPI_6_3_FIRMWARE_ACPI_CONTROL_STRUCTURE *) (UINTN) FadtTable->FirmwareCtrl;

```

Comment:
@mkopec does it collide somehow with these changes? https://github.com/Dasharo/edk2/pull/89/files

Reply:
@mkopec ping

Reply:
I don't think it collides, but it seems redundant, XFirmwareCtrl should always be non-null since https://github.com/Dasharo/coreboot/commit/a07b09ab71d04b870b9413b7e075a2369f59fd0a

Reply:
If S3 still works after these changes then it should be okay to merge

Reply:
I don't think "always" applies.  See https://github.com/Dasharo/coreboot/blob/a9d4e0ab871a7c71477315829ce22076c9435e31/src/acpi/acpi.c#L1060-L1088 (top and bottom parts).  It sets them, but then calls custom functions and [the one in `hudson/`](https://github.com/Dasharo/coreboot/blob/a9d4e0ab871a7c71477315829ce22076c9435e31/src/southbridge/amd/pi/hudson/fadt.c#L55-L56) does:
```c
	fadt->x_firmware_ctl_l = 0;	/* set to 0 if firmware_ctrl is used */
	fadt->x_firmware_ctl_h = 0;
```
It works without this change probably because something else updates the fields.

Reply:
It depends whether the FACS is above 4G or not. coreboot does not create the tables above 4G (maybe if long mode is enabled then perhaps yes). Anyways this is a good fallback if FACS 64bit address is zero, then use 32bit FACS address.